tests/FSharp.Core.UnitTests/FSharp.Core/Microsoft.FSharp.Core/NameOfTests.fs

```diff
@@ -0,0 +1,295 @@
+        let ``local property with encapsulated name and %.f`` = 0
+        let b = nameof(``local property with encapsulated name and %.f``)
+        Assert.AreEqual("local property with encapsulated name and %.f",b)
+
+[<TestFixture>]
+type MethodGroupNameOfTests() =
+    member this.MethodGroup() = ()    
+    member this.MethodGroup(i:int) = ()
+
+    member this.MethodGroup1(i:int, f:float, s:string) = 0
+    member this.MethodGroup1(f:float, l:int64) = "foo"
+    member this.MethodGroup1(u:unit -> unit -> int, h: unit) : unit = ()
+
+    [<Test>]
+    member this.``single argument method group name lookup`` () =
+        let b = nameof(this.MethodGroup)
+        Assert.AreEqual("MethodGroup",b)
+
+    [<Test>]
+    member this.``multiple argument method group name lookup`` () =
+        let b = nameof(this.MethodGroup1 : (float * int64 -> _))
+        Assert.AreEqual("MethodGroup1",b)
+
+[<TestFixture>]
+type FrameworkMethodTests() =
+    [<Test>]
+    member this.``library function name`` () =
+        let b = nameof(List.map)
+        Assert.AreEqual("map",b)
+
+    [<Test>]
+    member this.``static class function name`` () =
+        let b = nameof(System.Tuple.Create)
+        Assert.AreEqual("Create",b)
+
+type CustomUnionType =
+    | OptionA 
+    | OptionB of int * string
+
+type CustomRecordType =
+    { X: int; Y: int }
+
+[<Measure>] type Milliquacks 
+
+[<TestFixture>]
+type UnionAndRecordNameOfTests() =    
+
+    [<Test>]
+    member this.``measure 1`` () = 
+        let b = nameof(Milliquacks)
+        Assert.AreEqual("Milliquacks",b)
+
+    [<Test>]
+    member this.``record case 1`` () = 
+        let sample = Unchecked.defaultof<CustomRecordType>
+        let b = nameof(sample.X)
+        Assert.AreEqual("X",b)
+        let b = nameof(sample.Y)
+        Assert.AreEqual("Y",b)
+
+    [<Test>]
+    member this.``union case 1`` () = 
+        let b = nameof(OptionA)
+        Assert.AreEqual("OptionA",b)
+
+    [<Test>]
+    member this.``union case 2`` () = 
+        let b = nameof(OptionB)
+        Assert.AreEqual("OptionB",b)
+
+[<TestFixture>]
+type AttributeNameOfTests() =    
+
+    [<Test; System.Obsolete("test " + nameof(string))>]
+    member this.``ok in attribute`` () = 
+        let t = typeof<AttributeNameOfTests>.GetMethod("ok in attribute")
+        let attrs = t.GetCustomAttributes(typeof<ObsoleteAttribute>, false)
+        let attr = attrs.[0] :?> ObsoleteAttribute
+        Assert.AreEqual(attr.Message, "test string")
+
+[<TestFixture>]
+type OperatorNameOfTests() =    
+
+    [<Test>]
+    member this.``lookup name of typeof operator`` () =
+        let b = nameof(typeof<int>)
+        Assert.AreEqual("typeof",b)
+
+    [<Test>]
+    member this.``lookup name of + operator`` () =
+        let b = nameof(+)
+        Assert.AreEqual("op_Addition",b)
+
+    [<Test>]
+    member this.``lookup name of |> operator`` () =
+        let a = nameof(|>)
```

Comment:
Do we really want to allow `nameof` to get the actual names of the operators?

The purpose of `nameof` is to grab the name of a value(/variable), type or member that is represented in code.

Reply:
Yes I'm fine with this